### ISSUE TITLE
How to disable auto highlighting of all matching words

### ISSUE THREAD

**Issue by lafith** [syl20bnr/spacemacs#14880]
How to disable the following default highlighting of all matching words? I need to disable it globally, not just for a buffer. I prefer the normal cursor just like in vim.

![hh-2021-06-30_22 20 51](https://user-images.githubusercontent.com/39316548/124000947-bca1d400-d9f1-11eb-8437-58806ace35f2.gif)

**Comment by lebensterben** [comment]
what theme?
are you using LSP?

**Comment by korayal** [comment]
I think I was able to change to orange one via something like this:
```
  (set-face-attribute 'ahs-plugin-default-face nil
                      :background "#383838"
                      :foreground nil)
```
but I couldn't find the face used for other occurrences.

**Comment by lebensterben** [comment]
if this only appears in LSP, then you can turn off semantic highlighting.

**Comment by korayal** [comment]
@lebensterben turning off semantic highlighting doesn't affect this, I've already have it turned off. It's probably about something else.

**Comment by lebensterben** [comment]
there's also symbol highlighting you can turn off in LSP.

**Comment by korayal** [comment]
unfortunately `(setq lsp-enable-symbol-highlighting nil)` does not work for me, were there any other type of highlighting setting in LSP?

![image](https://user-images.githubusercontent.com/313801/124028688-f75d3980-d9fc-11eb-9ff5-e5b60497e05d.png)

**Comment by lebensterben** [comment]
what if you add auto-highlight-symbol to exclude packages?

**Comment by lafith** [comment]
> what theme?
> are you using LSP?

This is after the installation of spacemacs, no modifications. I have not included LSP and theme is the default one itself.

**Comment by lafith** [comment]
> I think I was able to change it to something less annoying via something like this:
> 
> ```
>   (set-face-attribute 'ahs-plugin-default-face nil
>                       :background "#3f3f3f"
>                       :foreground "orange")
> 
>   (set-face-attribute 'ahs-face nil
>                       :background "#383838"
>                       :foreground "Coral3")
> 
>   (set-face-attribute 'ahs-face-unfocused nil
>                       :background "#383838"
>                       :foreground "Coral3")
> ```
> 
> ![image](https://user-images.githubusercontent.com/313801/124026812-d0056d00-d9fa-11eb-81ae-25a515e6dabc.png)
> 
> @lebensterben yes, ~I think this is caused by LSP.~ looks like this is caused by something else.

This worked, thanks! It is less irritating with this combination of colors.



> what if you add auto-highlight-symbol to exclude packages?

This worked perfectly, no more highlighting of all matching occurences. Thanks!

**Comment by alexalekseyenko** [comment]
@lebensterben 
Unfortunately adding `auto-highlight-symbol` to `dotspacemacs-excluded-packages` changes functionality for `*` and `#` bindings.

Normally `*` and `#` are bound to `spacemacs/enter-ahs-forward` and `spacemacs/enter-ahs-backward`, respectively. When `auto-highlight-symbol` is removed, they are instead bound to `evil-ex-search-word-forward` and `evil-ex-search-word-backward`, the functionality of which is greatly reduced.

**Comment by lebensterben** [comment]
Umm so you want symbol highlighting only when you press * and #, but NOT when the cursor is hovering on a symbol? Am I right?

**Comment by alexalekseyenko** [comment]
Correct. Until recently that was the observed behavior, which is how I found this issue and, I assume, is why it was created in the first place.

Spacemacs uses this package for a number of its features:
https://github.com/syl20bnr/spacemacs/blob/532ad2567cba1d57d09e102c385315e7cfa829ec/layers/%2Bspacemacs/spacemacs-navigation/funcs.el#L91

I realize that this downstream dependency is not the responsibility of `auto-highlight-symbol`, but it would be nice to be able to control the new highlight behavior without losing existing functionality.

Additionally, LSP is increasingly the way people do things, and it would be nice to not clash with its own symbol highlighting functionality.

**Comment by alexalekseyenko** [comment]
I haven't had a chance to dig into it properly, but it looks like there were some pretty substantial changes upstream:
https://github.com/jcs-elpa/auto-highlight-symbol/commits/master/auto-highlight-symbol.el
particularly with respect to how `ahs-idle-timer` is being used, which previously allowed us to suppress the automatic highlighting behavior:
https://github.com/syl20bnr/spacemacs/blob/532ad2567cba1d57d09e102c385315e7cfa829ec/layers/%2Bspacemacs/spacemacs-navigation/packages.el#L73-L82

As I see it, we have three options:
- remove the `auto-highlight-symbol`-related functionality from Spacemacs entirely
- get upstream to either make or roll back a lot of changes
- use the version from last week

The first option seems like the worst one. The second option is possible, but may take some time if it happens at all. The third option costs us nothing and, AFAICT, will restore the desired functionality (which I thought was pretty great). Moreover, we can always merge the upstream changes in when we figure that out, and at worst, the first option is still an option, if not a good one.

What's the best way for us to pin an older version in our dependencies while we figure this out?

**Comment by bcc32** [comment]
@lafith, would you consider re-opening this issue so that it is more visible in searches?  I believe this issue should not be closed until Spacemacs's interaction with auto-highlight-symbol is fixed, since excluding the package is just a workaround, not a real fix, and it breaks manual highlighting when users press `*`/`#`.

**Comment by peaceful-james** [comment]
Thanks for all the helpful info everybody. The "orange text on orange background" is super annoying. I have not found a way to turn off the `auto-highlight-symbol-mode` in .spacemacs. I tried `(auto-highlight-symbol-mode -1)` in `user-config` but the mode is still enabled in every buffer.

In the end, I just added this to my .spacemacs (I use a dark theme and I wanted something that was almost like the highlight mode being disabled):

```
  ;; SYNTAX HIGHLIGHTING BS
  ;; https://github.com/syl20bnr/spacemacs/issues/14880
  ;; https://github.com/jcs-elpa/auto-highlight-symbol/issues/7
  (set-face-attribute 'ahs-plugin-default-face nil
                      :background "#1c1b1d"
                      :foreground "f5efef")

  (set-face-attribute 'ahs-face nil
                      :background "#1c1b1d"
                      :foreground "f5efef")

  (set-face-attribute 'ahs-face-unfocused nil
                      :background "#1c1b1d"
                      :foreground "f5efef")

  (set-face-attribute 'ahs-plugin-default-face-unfocused nil
                      :background "#1c1b1d"
                      :foreground "f5efef")
  )
```

**Comment by duianto** [comment]
This might work as a workaround until the main issue is resolved.

These two lines can be used to disable `auto-highlight-symbol-mode` from the user-config section.

```
  (remove-hook 'prog-mode-hook 'auto-highlight-symbol-mode)
  (remove-hook 'markdown-mode-hook 'auto-highlight-symbol-mode)
```

But it also disables the Spacemacs key bindings for opening/defining the 
`symbol highlight transient state`.

The following two keys revert to their default evil commands.
```
 *                      ;; evil-ex-search-word-forward
 #                      ;; evil-ex-search-word-backward
```
They don't jump by symbol, but by the word under the cursor.

**Comment by duianto** [comment]
A PR is pending with a possible fix.
Fix symbol highlighting #14892

There are still some issues with toggling the symbol highlighting, but the other fixes probably are better than the current state.

**Comment by duianto** [comment]
An issue has been opened in the `spacemacs-theme` repository,
about the evil normal state cursor disappearing against the `auto-highlight-symbol`'s default color.
auto-highlight-symbol, new default face, hard to see cursor https://github.com/nashamri/spacemacs-theme/issues/186

![emacs_l18w2yvvMe](https://user-images.githubusercontent.com/13420573/124361823-6f13aa00-dc31-11eb-8f23-13dde16fdf31.png)

**Comment by alexalekseyenko** [comment]
Why is it that we can't just pin an older version of `auto-highlight-symbol`? It seemed to be working for years before the most recent changes.

**Comment by lebensterben** [comment]
the issue is not caused by using latest release of packages. The real issue is Emacs package ecosystem doesn't support something like semantic versioning which can avoid lots of breaking changes.

**Comment by practicalli-john** [comment]
I just updated packages today and have automatic highlighting enabled by default.  My temporary hack is to move the cursor so no matches are highlighted and use `SPC SPC ahs-change-range` which should change the plugin to display area.

Unfortunately this needs to be done for each buffer, but seems to survive opening a file again (havent tested it after restarting Emacs).

I am using doom-gruvbox and doom-grubbox-light themes.

**Comment by duianto** [comment]
Does the current workaround mentioned above work for you?
https://github.com/syl20bnr/spacemacs/issues/14880#issuecomment-873395165

Add this to the user-config section.
```
  (remove-hook 'prog-mode-hook 'auto-highlight-symbol-mode)
  (remove-hook 'markdown-mode-hook 'auto-highlight-symbol-mode)
```

There's a pending PR with some symbol highlight toggling issues:
Fix symbol highlighting #14892

**Comment by practicalli-john** [comment]
@duianto my preferred resolution would be for a PR that makes the `(spacemacs/toggle-automatic-symbol-highlight)` command work with the new changes (I appreciate this would take some work)

Removing hooks would disable the keybindings of the Spacemacs Symbol Highlighting Transient state, that would take away functionality I have traditionally used.  I will evaluate if I still need these key bindings though

I may be okay with just LSP and iedit highlighting, which I suspect is using a different face to the auto-highlighting face.  If that is the case, then removing the autohighlighting package and features would seem the most efficient solution (if nothing else breaks)

The lsp highlighting for code with the doom-grubox-theme was quite harsh, although the maintainer kindly [changed the face to be something much more subtle](https://github.com/hlissner/emacs-doom-themes/pull/580#issuecomment-851049815).  

It seems strange to resolve the auto highlighting issue by just hiding the face on every theme though, but can understand this is a relevant quick fix for some people.

The most relevant fix for myself is to use the old package until I evaluate if I need the Spacemacs symbol evaluation.  If not, then I will remove all hooks for that mode.

**Comment by duianto** [comment]
I've pushed an update to the PR:
Fix symbol highlighting #14892

It's almost working. Details in the comment: https://github.com/syl20bnr/spacemacs/pull/14892#issuecomment-875151413

**Comment by duianto** [comment]
The PR has been merged, the symbol highlighting is now disabled on startup.

@lafith Could you pull the latest develop branch and confirm that it works as expected. If it does then you can close this issue.
